src/verso/Verso/Code/HighlightedToTex.lean

```diff
@@ -118,45 +114,59 @@ public defmethod Highlighted.toVerbatimTeX : Highlighted → Verso.Output.TeX
 
-public defmethod Highlighting.Token.toTeX (t : Highlighting.Token) : Verso.Output.TeX :=
-  verbatim (t.toVerbatimTeX)
+public defmethod Highlighting.Token.toTeX [Monad m] [GenreTeX g m] (t : Highlighting.Token) :
+    TeXT g m Verso.Output.TeX :=
+  verbatimInline (t.toVerbatimTeX)
 
-public defmethod Highlighted.toTeX (t : Highlighted) : Verso.Output.TeX :=
-  let strip := t.trimOneTrailingNl
+public defmethod Highlighted.toTeX [Monad m] [GenreTeX g m] (t : Highlighted) :
+    TeXT g m Verso.Output.TeX :=
+  let strip := t.trimOneTrailingNl.trimOneLeadingNl
   if strip.isEmpty then
-    .empty
+    pure .empty
+  else if strip.containsNewline then
+    pure <| verbatimBlock (strip.toVerbatimTeX)
   else
-    verbatim (strip.toVerbatimTeX)
+    verbatimInline (strip.toVerbatimTeX)
+
 
 open Verso.Output.TeX in
-public defmethod Highlighted.Goal.toTeX (h : Highlighted.Goal Highlighted) : Id Verso.Output.TeX := do
+public defmethod Highlighted.Goal.toTeX [Monad m] [GenreTeX g m] (h : Highlighted.Goal Highlighted) : TeXT g m Verso.Output.TeX := do
   let {name, goalPrefix, hypotheses, conclusion} := h
   let mut rows : Array TeX := #[]
+  let verbatim (t : Verso.Output.TeX) : Verso.Output.TeX :=
+    .seq #[.raw "\\LeanVerb|", t, .raw "|"]
   if let some n := name then
     rows := rows ++ #[\TeX{\textbf{"case"} " " \Lean{n}}]
-  let toRow (h : Highlighted.Hypothesis Highlighted) : TeX :=
-    let namesTeX := h.names.map (·.toTeX) |>.toList.intersperse (.text " ")
-    .seq #[namesTeX, .text " : ", h.typeAndVal.toTeX]
-  rows := rows ++ hypotheses.map toRow
-  rows := rows ++ #[verbatim (goalPrefix) ++ conclusion.toTeX]
-  \TeX{\begin{tabular}{"l"} \Lean{rows.map (· ++ .raw "\\\\")} \end{tabular}}
-
-def messageContentsToVerbatimTeX (h : Highlighted.MessageContents Highlighted) : Verso.Output.TeX :=
-  match h with
-  | .text str => str
-  | .goal g => g.toTeX -- FIXME: this doesn't seem correct?
-  | .term t => t.toVerbatimTeX
-  | .trace _cls _msg _children _collapsed => .empty -- FIXME: what to render here?
-  | .append mcs => .seq (mcs.map messageContentsToVerbatimTeX)
 
-public defmethod Highlighted.Message.toTeX (h : Highlighted.Message) : Verso.Output.TeX :=
+  let toRow (h : Highlighted.Hypothesis Highlighted) : TeXT g m Verso.Output.TeX  := do
+    let namesTeX := (← h.names.mapM (·.toTeX)).toList.intersperse (.text " ")
+    pure <| .seq #[namesTeX, .text " : ", (← h.typeAndVal.toTeX)]
+  rows := rows ++ (← hypotheses.mapM toRow)
+  rows := rows ++ #[verbatim (goalPrefix) ++ (← conclusion.toTeX)]
+  pure \TeX{\begin{tabular}{"l"} \Lean{rows.map (· ++ .raw "\\\\")} \end{tabular}}
+
+def messageContentsToVerbatimTeX [Monad m] [GenreTeX g m] (h : Highlighted.MessageContents Highlighted) : TeXT g m Verso.Output.TeX :=
+  match h with
+  | .text str => pure str
+  | .goal g => pure g.toString
+  | .term t => pure t.toVerbatimTeX
+  | .trace _cls _msg _children _collapsed => pure .empty -- FIXME: what to render here?
```

Comment:
For TeX, the results should show the children if `collapsed` is true. I think the goal should be to match how it looks in the InfoView reasonably well. Maybe some kind of nested tabular or nested custom list-like environment?

It should probably always show `[cls] msg`, and then if `collapsed = false`, show the children underneath, nested and indented.